src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -665,6 +665,8 @@ private function validateAndDenormalize(Type $type, string $currentClass, string
             if ($t instanceof CollectionType) {
                 $collectionKeyType = $t->getCollectionKeyType();
                 $collectionValueType = $t->getCollectionValueType();
+            } elseif ($t instanceof Type\TemplateType) {
```

Comment:
That condition must IMO go under the first `if` of the `foreach` (L663).
Otherwise, a template of collection won't be handled properly.

Reply:
It seems to handle collections (i.e. `T[]`) properly this has to be done inside the `if ($t instanceof CollectionType)` block

```php
                if ($collectionValueType instanceof TemplateType) {
                    $collectionValueType = $collectionValueType->getBound();
                }
``` 

_(maybe same should be for `$collectionValueKey` too)_

I wonder would it make more sense to make `TemplateType::getBaseType()` return `$this->getBound()` instead of throwing – then we didn't need any of this here. I'm not too familiar with the new TypeInfo component so this could be a silly proposal.

Reply:
> I wonder would it make more sense to make TemplateType::getBaseType() return $this->getBound() instead of throwing – then we didn't need any of this here. I'm not too familiar with the new TypeInfo component so this could be a silly proposal.

I indeed think that the bound type should be considered as the base type. Could you please send a PR in that way?

Reply:
Sure, [here](https://github.com/symfony/symfony/pull/58259)